gateway_code/nodes.py

```diff
@@ -0,0 +1,192 @@
+# with a limited warranty  and the software's author,  the holder of the
+# economic rights,  and the successive licensors  have only  limited
+# liability.
+#
+# The fact that you are presently reading this means that you have had
+# knowledge of the CeCILL license and that you accept its terms.
+
+""" Common logic for plugin nodes classes """
+import abc
+import os
+import pkgutil
+
+from gateway_code.utils import elftarget
+
+
+class MetaNodeBase(type):
+    """ Base Metaclass for Plugin Control or Open Nodes"""
+
+    def __init__(cls, name, bases, class_dict):
+        type.__init__(cls, name, bases, class_dict)
+        if not hasattr(cls, '__registry__'):
+            cls.__registry__ = {}
+        else:
+            cls.__registry__[cls.TYPE] = cls
+
+
+class MetaNode(MetaNodeBase, abc.ABCMeta):
+    """ Combining metaclasses,
+    control or open Nodes have a contract verified on instantiation """
+    pass
+
+
+class ControlNode(object):
+    """ Class to inherit, for control node classes """
+    __metaclass__ = MetaNode
+    TYPE = 'undefined'
+
+    @abc.abstractmethod
+    def start(self, exp_id, exp_files=None):
+        """ This method is called when starting an experiment """
+        pass
+
+    @abc.abstractmethod
+    def stop(self):
+        """ This method is called when stopping an experiment """
+        pass
+
+    @abc.abstractmethod
+    def start_experiment(self, profile):
+        """ Configure experiment and monitoring on ControlNode"""
+        pass
+
+    @abc.abstractmethod
+    def stop_experiment(self):
+        """ Cleanup Control node Monitoring and experiment """
+        pass
+
+    @abc.abstractmethod
+    def autotest_setup(self, measures_handler):
+        """ Setup the control node for the open node autotest """
+        pass
+
+    @abc.abstractmethod
+    def autotest_teardown(self, stop_on):
+        """ Cleanup the control node for the open node autotest """
+        pass
+
+    @abc.abstractmethod
+    def configure_profile(self, profile=None):
+        """ Setup the profile used by the control node """
+        pass
+
+    @abc.abstractmethod
+    def status(self):
+        """ Status of the control node """
+        pass
+
+    @abc.abstractmethod
+    def flash(self, firmware_path):
+        """ Flash firmware on the control_node """
+        pass
+
+    @property
+    def tty(self):
+        """ The node tty """
+        return '/dev/ttyCN_%s' % self.TYPE
+
+
+class OpenNode(object):
+    """ class to inherit, for open node classes """
+    __metaclass__ = MetaNode
+    TYPE = 'undefined'
+
+    @abc.abstractmethod
+    def setup(self, firmware_path):
+        """ Setup the open node with a firmware"""
+        pass
+
+    @abc.abstractmethod
+    def teardown(self):
+        """ Cleanup the open node """
+        pass
+
+    @abc.abstractmethod
+    def status(self):
+        """ Status of the open node """
+        pass
+
+    @property
```

Comment:
I think this method could be shared in the NodeBase class ?

Reply:
OK, I've added a NodeBase class with the status method that is common between control and open nodes in it, that should remove that code duplication